File: src/parport/ParPort.cpp
//
//  libavg - Media Playback Engine. 
//  Copyright (C) 2003-2006 Ulrich von Zadow
//
//  This library is free software; you can redistribute it and/or
//  modify it under the terms of the GNU Lesser General Public
//  License as published by the Free Software Foundation; either
//  version 2 of the License, or (at your option) any later version.
//
//  This library is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
//  Lesser General Public License for more details.
//
//  You should have received a copy of the GNU Lesser General Public
//  License along with this library; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//
//  Current versions can be found at www.libavg.de
//

#include "ParPort.h"

#include "../avgconfigwrapper.h"
#include "../base/Logger.h"
#include "../base/Exception.h"
#include "../base/MathHelper.h"

#ifdef AVG_ENABLE_PARPORT 
#include <linux/ppdev.h>
#include <linux/parport.h>
#endif

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sys/ioctl.h>

#include <cstring>
#include <iostream>
#include <sstream>
#include <cerrno>

using namespace std;

namespace avg {

ParPort::ParPort()
    : _myFileDescriptor(-1),
      _myDeviceName(""),
      _isOpen(false),
      _myDataLines(0)
{
}

ParPort::~ParPort() {
    deinit();
}

void 
ParPort::init(const string& theDevice) {
    string myDevice = theDevice;
#ifdef AVG_ENABLE_PARPORT
    if (myDevice.empty()) {
        myDevice = "/dev/parport0";
    }
    _myFileDescriptor = open(myDevice.c_str(), O_RDONLY);
    if (_myFileDescriptor == -1) {
        AVG_TRACE(Logger::ERROR, 
                "Failed to open parallel port '" 
                 << myDevice << "': " << strerror(errno));
        return;
    } else {
        AVG_TRACE(Logger::CONFIG, "Parallel port opened.");
    }

    if (ioctl(_myFileDescriptor, PPCLAIM) == -1) {
        AVG_TRACE(Logger::ERROR,
                "Failed to claim parallel port: "
                 << strerror(errno));
        _myFileDescriptor = -1;
        return;
    }
    _isOpen = true;
#else
    AVG_TRACE(Logger::ERROR,
            "Failed to open parallel port. Support not compiled in.");
    _myFileDescriptor = -1;
#endif
    _myDeviceName = myDevice;
}

bool ParPort::setControlLine(int theLine, bool theValue) {
    int theStatus;
    if (theValue) {
        theStatus = theLine;
    } else {
        theStatus = 0;
    }
    return frob(theLine, theStatus); 
}

bool ParPort::getStatusLine(int theLine) {
#ifdef AVG_ENABLE_PARPORT
    if (_myFileDescriptor == -1) {
        return false;
    }
    unsigned char myStatus;
    int myOk = ioctl(_myFileDescriptor, PPRSTATUS, &myStatus);
    if (myOk == -1) {
        AVG_TRACE(Logger::ERROR,
                "Could not get parallel port status.");
        return false;
    }
    return (myStatus & theLine) == theLine;
#else
    return false;
#endif
}

bool ParPort::setDataLines(unsigned char theData)
{
#ifdef AVG_ENABLE_PARPORT
    if (_myFileDescriptor == -1) {
        return false;
    }
    _myDataLines |= theData;
    int myOk = ioctl(_myFileDescriptor, PPWDATA, &_myDataLines);
    if (myOk == -1) {
        AVG_TRACE(Logger::ERROR,
                "Could not write parallel port data.");
        return false;
    }    
    return true;
#else
    return false;
#endif
}

bool ParPort::clearDataLines(unsigned char theData)
{
#ifdef AVG_ENABLE_PARPORT
    if (_myFileDescriptor == -1) {
        return false;
    }
    _myDataLines &= ~theData;
    int myOk = ioctl(_myFileDescriptor, PPWDATA, &_myDataLines);
    if (myOk == -1) {
        AVG_TRACE(Logger::ERROR,
                "Could not write parallel port data.");
        return false;
    }    
    return true;
#else
    return false;
#endif
}

bool ParPort::setAllDataLines(unsigned char theData)
{
#ifdef AVG_ENABLE_PARPORT
    if (_myFileDescriptor == -1) {
        return false;
    }
    _myDataLines = theData;
    int myOk = ioctl(_myFileDescriptor, PPWDATA, &_myDataLines);
    if (myOk == -1) {
        AVG_TRACE(Logger::ERROR,
                "Could not write parallel port data.");
        return false;
    }    
    return true;
#else
    return false;
#endif
}    

bool ParPort::isAvailable() 
{
    return (_myFileDescriptor != -1);
}

bool ParPort::frob(int theLines, int theStatus) 
{
#ifdef AVG_ENABLE_PARPORT
    if (_myFileDescriptor == -1) {
        return false;
    }
    ppdev_frob_struct myFrob;
    myFrob.mask = theLines;
    myFrob.val = theStatus;
    int myOk = ioctl(_myFileDescriptor, PPFCONTROL, &myFrob);
    if (myOk == -1) {
        AVG_TRACE(Logger::ERROR,
                "Could not set parallel port control line.");
        return false;
    }
    return true;
#else
    return false;
#endif
}

void ParPort::deinit() {
    if (_isOpen) {
        if (::close(_myFileDescriptor) == -1) {
            AVG_TRACE(Logger::ERROR,
                    "Can't close parallel port '"
                    << _myDeviceName << "':" << strerror(errno));
        } else {
            AVG_TRACE(Logger::CONFIG, "Parallel port closed.");
        }
    }
}

bool
ParPort::writeControlRegister(unsigned char theStatus) {
#ifdef AVG_ENABLE_PARPORT
    if (_isOpen) {
        if (ioctl(_myFileDescriptor, PPWCONTROL, & theStatus) == -1) {
            AVG_TRACE(Logger::ERROR,
                    "ERROR: Failed to write control register: " 
                    << strerror(errno));
            return false;
        }
        return true;
    }
#endif
    return false;
}

}

File: src/test/parport.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# libavg - Media Playback Engine.
# Copyright (C) 2003-2008 Ulrich von Zadow
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
# Current versions can be found at www.libavg.de
#

import unittest

import sys
sys.path.append('/usr/local/lib/python2.3/site-packages/avg')
import avg

def keyup():
    def printPPLine(line, name):
        print name,
        if ParPort.getStatusLine(line):
            print ": off",
        else:
            print ":  on",
    Event = Player.getCurEvent()
    print Event.keystring
    if Event.keystring == "1":
        ParPort.setDataLines(avg.PARPORTDATA1)
    elif Event.keystring == "2":
        ParPort.setDataLines(avg.PARPORTDATA2)
    elif Event.keystring == "3":
        ParPort.setDataLines(avg.PARPORTDATA3)
    elif Event.keystring == "4":
        ParPort.setDataLines(avg.PARPORTDATA4)
    elif Event.keystring == "5":
        ParPort.setDataLines(avg.PARPORTDATA5)
    elif Event.keystring == "6":
        ParPort.setDataLines(avg.PARPORTDATA6)
    elif Event.keystring == "7":
        ParPort.setDataLines(avg.PARPORTDATA7)
    elif Event.keystring == "8":
        ParPort.setDataLines(avg.PARPORTDATA0)
    elif Event.keystring == "q":
        ParPort.clearDataLines(avg.PARPORTDATA1)
    elif Event.keystring == "w":
        ParPort.clearDataLines(avg.PARPORTDATA2)
    elif Event.keystring == "e":
        ParPort.clearDataLines(avg.PARPORTDATA3)
    elif Event.keystring == "r":
        ParPort.clearDataLines(avg.PARPORTDATA4)
    elif Event.keystring == "t":
        ParPort.clearDataLines(avg.PARPORTDATA5)
    elif Event.keystring == "z":
        ParPort.clearDataLines(avg.PARPORTDATA6)
    elif Event.keystring == "u":
        ParPort.clearDataLines(avg.PARPORTDATA7)
    elif Event.keystring == "i":
        ParPort.clearDataLines(avg.PARPORTDATA0)
    elif Event.keystring == "a":
        ParPort.setControlLine(avg.CONTROL_STROBE, 1)
    elif Event.keystring == "s":
        ParPort.setControlLine(avg.CONTROL_AUTOFD, 1)
    elif Event.keystring == "f":
        ParPort.setControlLine(avg.CONTROL_SELECT, 1)
    elif Event.keystring == "d":
        ParPort.setControlLine(avg.CONTROL_INIT, 1)
    elif Event.keystring == "y":
        ParPort.setControlLine(avg.CONTROL_STROBE, 0)
    elif Event.keystring == "x":
        ParPort.setControlLine(avg.CONTROL_AUTOFD, 0)
    elif Event.keystring == "c":
        ParPort.setControlLine(avg.CONTROL_SELECT, 0)
    elif Event.keystring == "v":
        ParPort.setControlLine(avg.CONTROL_INIT, 0)
    printPPLine (avg.STATUS_ERROR, "STATUS_ERROR")
    print ", ",
    printPPLine (avg.STATUS_SELECT, "STATUS_SELECT")
    print ", ",
    printPPLine (avg.STATUS_PAPEROUT, "STATUS_PAPEROUT")
    print ", ",
    printPPLine (avg.STATUS_ACK, "STATUS_ACK")
    print ", ",
    printPPLine (avg.STATUS_BUSY, "STATUS_BUSY")
    print

ParPort = avg.ParPort()
ParPort.init("")

Player = avg.Player.get()
Player.loadFile("parport.avg")
Player.play(30)

